src/core/triton_repo_agent.h

```diff
@@ -0,0 +1,131 @@
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include "src/core/tritonserver_apis.h"
+
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+#include <vector>
+#include "src/core/constants.h"
+#include "src/core/filesystem.h"
+#include "src/core/model_config_utils.h"
+
+namespace nvidia { namespace inferenceserver {
+
+class TritonRepoAgent {
+ public:
+  using Parameters = std::vector<std::pair<std::string, std::string>>;
+  typedef TRITONSERVER_Error* (*TritonRepoAgentInitFn_t)(
+      TRITONREPOAGENT_Agent* agent);
+  typedef TRITONSERVER_Error* (*TritonRepoAgentFiniFn_t)(
+      TRITONREPOAGENT_Agent* agent);
+  typedef TRITONSERVER_Error* (*TritonRepoAgentModelActionFn_t)(
+      TRITONREPOAGENT_Agent* agent, TRITONREPOAGENT_AgentModel* model,
+      const TRITONREPOAGENT_ActionType action_type);
+
+  static Status Create(
+      const std::string& name, const std::string& libpath,
+      std::shared_ptr<TritonRepoAgent>* agent);
+  ~TritonRepoAgent();
+
+  TritonRepoAgentModelActionFn_t AgentModelActionFn() const
+  {
+    return model_action_fn_;
+  }
+
+ private:
+  TritonRepoAgent(const std::string& name)
+      : name_(name), dlhandle_(nullptr), init_fn_(nullptr), fini_fn_(nullptr),
+        model_action_fn_(nullptr)
+  {
+  }
+  const std::string name_;
+
+  // dlopen / dlsym handles
+  void* dlhandle_;
+  TritonRepoAgentInitFn_t init_fn_;
+  TritonRepoAgentFiniFn_t fini_fn_;
+  TritonRepoAgentModelActionFn_t model_action_fn_;
+};
+
+class TritonRepoAgentModel {
```

Comment:
This should represent a single agent and corresponds to TRITONREPOAGENT_AgentModel. It should hold the state associated with the AgentModel (and the TRITONREPOAGENT API needs to be enhanced with SetState and State methods. The idea is that this AgentModel object persists as long as the model is loaded... Instead of having Initialize/Finalize the TRITONREPOAGENT_ModelAction will indicate its life-cycle. LOAD when it is created and UNLOAD indicates it is about to be destroyed (we need to add the UNLOAD).

Reply:
We'll need something in the model repository manager to hold the list of these associated with a model.

Reply:
I was thinking that the implementation of `TRITONREPOAGENT_AgentModel` would be holding the list of agents associated with the model and the specific parameters to be passed to the agents. In this way the model repository manager would be holding one "agent model" for one model.

From your description, it seems like you are expecting a "agent model" to be associated with only one agent and the model repository manager will hold a list of "agent model"s for one model?

I think both ideas are the same, just being represented differently

Reply:
Within Triton core it may not matter so much but for the person writing an agent the AgentModel abstraction represents just one model using that agent (with associated user state). Since it needs to be that way for person writing the agent I'd prefer the abstractions within Triton to be the same.